lib/SourceMapDevToolPlugin.js

```diff
@@ -43,57 +74,43 @@ class SourceMapDevToolPlugin {
 		const requestShortener = new RequestShortener(compiler.context);
 		const options = this.options;
 		options.test = options.test || /\.(js|css)($|\?)/i;
+
+		const matchObject = ModuleFilenameHelpers.matchObject.bind(undefined, options);
+
 		compiler.plugin("compilation", compilation => {
 			new SourceMapDevToolModuleOptionsPlugin(options).apply(compilation);
+
 			compilation.plugin("after-optimize-chunk-assets", function(chunks) {
 				const moduleToSourceNameMapping = new Map();
-				const tasks = [];
+				let tasks = [];
+
 				chunks.forEach(function(chunk) {
-					chunk.files.filter(ModuleFilenameHelpers.matchObject.bind(undefined, options)).map(function(file) {
-						const asset = compilation.assets[file];
-						if(asset.__SourceMapDevToolFile === file && asset.__SourceMapDevToolData) {
-							const data = asset.__SourceMapDevToolData;
-							for(const cachedFile in data) {
-								compilation.assets[cachedFile] = data[cachedFile];
-								if(cachedFile !== file)
-									chunk.files.push(cachedFile);
-							}
-							return;
-						}
-						let source, sourceMap;
-						if(asset.sourceAndMap) {
-							const sourceAndMap = asset.sourceAndMap(options);
-							sourceMap = sourceAndMap.map;
-							source = sourceAndMap.source;
-						} else {
-							sourceMap = asset.map(options);
-							source = asset.source();
-						}
-						if(sourceMap) {
-							return {
-								chunk,
-								file,
-								asset,
-								source,
-								sourceMap
-							};
-						}
-					}).filter(Boolean).map(task => {
-						const modules = task.sourceMap.sources.map(source => {
-							const module = compilation.findModule(source);
-							return module || source;
-						});
-						for(const module of modules) {
-							if(!moduleToSourceNameMapping.get(module)) {
-								moduleToSourceNameMapping.set(module, ModuleFilenameHelpers.createFilename(module, moduleFilenameTemplate, requestShortener));
+					tasks = chunk.files.reduce((queue, file) => {
```

Comment:
btw. this part of the PR is covered by tests. Only the caching thing above is not covered. 

btw2. I can recommend the Codecov Extension which highlights covered lines...